design/kubebuilder-migration.md

```diff
@@ -0,0 +1,64 @@
+# Move Contour to use kubebuilder instead of informers
+
+Status: Draft
+
+Contour's method of interaction with Kubernetes is two years old, and the community has moved on from Informers.
+This proposal suggests that we should move to using Kubebuilder [controller-runtime][1] for interacting with Kubernetes.
+This will make extending Contour with additional CRDs as we need to more straightforward.
+
+## Goals
+
+- Migrate Contour to use Kubebuilder [controller-runtime][1] instead of informers.
+
+## Non Goals
+
+- Change anything after the DAG builder
+- Implement the new service-api types as part of this proposal
+
+## Background
+
+Two years ago, @davecheney built Contour to use custom-wired code to talk to Kubernetes.
+At the time, this was a sensible decision, as there were no well-developed frameworks or common methods for interacting with Kubernetes.
+Since then, the community has done a large amount of work to simplify the construction of controllers, currently embodied in the kubebuilder [controller-runtime][1] library.
+
+This has become an issue now because we are looking at adding support for further CRDs, (the SIG-Network service-apis) and want to leave the door open for more if we need to.
+
+In addition, moving to using controller-runtime means that we have access to frameworks for helping with:
+
+- leader election
+- defaulting values for CRDs
+- webhooks to translate between CRD versions (related to #2068)
+
+without having to custom-build that code.
+
+Lastly, this will allow us to consume the new service-api types and any other CRDs (including our own), without generating the client-go code to make Informers.
+The service-api project currently does not generate these.
+
+This is a pretty big refactor though, so we should agree it's a good idea before starting.
```

Comment:
probably not appropriate commentary.

It is probably worth pointing out in the high level design that because of our desire to avoid a long running feature branch and a with make sure the product is always shippable, the incremental approach is the way to go.